napari/_vispy/vispy_axes_visual.py

```diff
@@ -131,51 +199,94 @@ def __init__(self, axes, camera, dims, parent=None, order=0):
+        reversed_axes = [self._dims.ndim - 1 - a for a in axes[::-1]]
+
+        # Determine colors of axes based on reverse position
+        if self._axes.colored:
+            axes_colors = [
+                self._default_color[ra % len(self._default_color)]
+                for ra in reversed_axes
+            ]
+        else:
             color = np.subtract(1, self._axes.background_color)[:3]
-            arrow_color = [color] * self._NUM_SEGMENTS_ARROWHEAD * 3
+            axes_colors = [color] * self._dims.ndisplay
+
+        # Make sure have enough colors and labels for displayed dimensions
+        if len(axes_colors) < self._dims.ndisplay:
+            axes_colors += [self._axes.background_color] * (
+                self._dims.ndisplay - len(axes_colors)
+            )
+        if len(axes_labels) < self._dims.ndisplay:
+            axes_labels += [''] * (self._dims.ndisplay - len(axes_labels))
+
+        # Determine data based on number of displayed dimensions and
+        # axes visualization parameters
+        if self._axes.dashed and self._dims.ndisplay == 2:
+            data = self._dashed_line_data2D
+            color = color_dashed_lines(axes_colors)
+            text_data = self._line_data2D[1::2]
+        elif self._axes.dashed and self._dims.ndisplay == 3:
+            data = self._dashed_line_data3D
+            color = color_dashed_lines(axes_colors)
+            text_data = self._line_data3D[1::2]
+        elif not self._axes.dashed and self._dims.ndisplay == 2:
+            data = self._line_data2D
+            color = color_lines(axes_colors)
+            text_data = self._line_data2D[1::2]
+        elif not self._axes.dashed and self._dims.ndisplay == 3:
+            data = self._line_data3D
+            color = color_lines(axes_colors)
+            text_data = self._line_data3D[1::2]
         else:
-            if self._axes.dashed:
-                color = self._dashed_color
-                arrow_color = self._default_arrow_color
-            else:
-                color = self._default_color
-                arrow_color = self._default_arrow_color
-
-        if self._axes.arrows:
-            arrow_vertices = self._default_arrow_vertices
-            arrow_faces = self._default_arrow_faces
+            raise ValueError(
+                'Axes dash status and ndisplay combination not supported'
+            )
+
+        if self._axes.arrows and self._dims.ndisplay == 2:
+            arrow_vertices = self._default_arrow_vertices2D
+            arrow_faces = self._default_arrow_faces2D
+            arrow_color = color_arrowheads(
+                axes_colors, self._NUM_SEGMENTS_ARROWHEAD
+            )
+        elif self._axes.arrows and self._dims.ndisplay == 3:
+            arrow_vertices = self._default_arrow_vertices3D
+            arrow_faces = self._default_arrow_faces3D
+            arrow_color = color_arrowheads(
+                axes_colors, self._NUM_SEGMENTS_ARROWHEAD
+            )
```

Comment:
This is a mess and looks very much ripe for simplification. I might suggest a dictionary mapping a tuple of (ndisplay, dashed, arrows) to a set of vertices, colors, and data. e.g.:


```python
# in e.g. constructor
self._axes_elements = {
    (2, True, False): (self._dashed_line_data2D, color_dashed_lines(axes_colors), None, None),
    ...
}

# ... then:

data, colors, arrow_vertices, arrow_faces = (
        self._axes_elements[(self._dims.ndisplay, self._axes.dashed, self._axes.arrows)]
        )

text_data = data[1::2]
arrow_color = color_arrowheads(axes_colors, self._NUM_SEGMENTS_ARROWHEAD)
```

I consider this optional before merge, but I'll just note that there's lots of bits in the codebase that we are still waiting to clean up with this particular if/else cascade pattern...

Reply:
(Also, there might be a cleaner pattern than a dict that I'm not seeing yet.)

Reply:
so i really like this idea, but i can't get it to work without more if statements, on problem was `axes_colors` gets calculate in the update, so the dict in the constructor needs to return a function, that was solvable, but the arrow coloring needed an extra if statement, and the text_data wasn't quite so simple either.

I think we can revisit this in a future PR, but thanks for the suggestion